File: src/common/flow_import_specifier.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type userland = string [@@deriving show, eq, ord]

let userland x = x

let map_userland ~f x = f x

let display_userland x = x

let unwrap_userland x = x

module Key = struct
  type t =
    | Userland of userland
    | HasteImportWithSpecifiedNamespace of {
        namespace: Bitset.t;
        name: string;
      }
  [@@deriving show, ord]
end

let userland_specifier x = Key.Userland x

module Map = WrappedMap.Make (Key)
module Set = Flow_set.Make (Key)
include Key

File: rust_port/crates/flow_common/src/flow_import_specifier.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;

#[derive(
    Debug,
    Clone,
    Dupe,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    serde::Serialize,
    serde::Deserialize
)]
pub struct Userland(FlowSmolStr);

impl Userland {
    pub fn from_smol_str(x: FlowSmolStr) -> Self {
        Self(x)
    }

    pub fn map(&self, f: impl FnOnce(&str) -> FlowSmolStr) -> Self {
        Self(f(&self.0))
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn into_inner(self) -> FlowSmolStr {
        self.0
    }

    pub fn display(&self) -> &str {
        &self.0
    }
}

#[derive(
    Debug,
    Clone,
    Dupe,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    serde::Serialize,
    serde::Deserialize
)]
pub enum FlowImportSpecifier {
    Userland(Userland),
}

impl FlowImportSpecifier {
    pub fn userland(x: FlowSmolStr) -> Self {
        Self::Userland(Userland(x))
    }
}
